### PR TITLE
Implement Application management

### LINKED ISSUE #5: Manage an Application Instance
If a manifest includes an 'Application' we should be able to manage it (name, labels, GVKs, assembly phase, ...)

### PR DESCRIPTION
**Pull request by johnsonj**
The Application is a CR that represents a given deployment. If a
deployment contains an Application we can specialize it based on the
declarative configuration. We can go further with Addons which have
health/version to expose on the Application.

More info on the application: https://github.com/kubernetes-sigs/application

fixes #5

### CODE REVIEW

File: examples/dashboard-operator/channels/packages/dashboard/1.8.3/manifest.yaml [hunk 1]
```diff
@@ -165,3 +165,28 @@ spec:
...
+  namespace: kube-system
+spec:
+  descriptor:
+    type: "kubernetes-dashboard"
+    description: "Kubernetes Dashboard is a general purpose, web-based UI for Kubernetes clusters. It allows users to manage applications running in the cluster and troubleshoot them, as well as manage the cluster itself."
+    icons:
+    - src: "https://github.com/kubernetes/kubernetes/raw/master/logo/logo.png"
+      type: "image/png"
+    maintainers:
+    - name: Jeff Johnson
```
Comment (justinsb) [inline examples/dashboard-operator/channels/packages/dashboard/1.8.3/manifest.yaml]: You might want to say here that this is a demo app, lest people think this is something you want people to run in production!

File: pkg/patterns/addon/application.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	if !ok {
+		return fmt.Errorf("instance %T was not an addonsv1alpha1.CommonObject", instance)
+	}
+
+	var app *manifest.Object
+	for _, o := range objects.Items {
+		if o.Group == "app.k8s.io" && o.Kind == "Application" {
+			if app != nil {
+				log.Info("cannot update application with multiple app.k8s.io/Application in manifest")
+				return nil
```
Comment (justinsb) [inline pkg/patterns/addon/application.go]: Should this be an error?

File: pkg/patterns/addon/application.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+			}
+			app = o
+		}
+	}
+
+	if app == nil {
+		log.Info("cannot transformApplication without an app.k8s.io/Application in the manifest")
+		return nil
```
Comment (justinsb) [inline pkg/patterns/addon/application.go]: Maybe an error here also?

File: pkg/patterns/addon/application.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	}
+
+	assemblyPhase := Pending
+	if addonObject.GetCommonStatus().Healthy {
+		assemblyPhase = Succeeded
+	}
+
+	// TODO: Version should be on CommonStatus as well
```
Comment (justinsb) [inline pkg/patterns/addon/application.go]: Right - and probably on the app status also.  I see the difference as desired-version vs current-installed-version

File: pkg/patterns/declarative/application.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+	"sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/kubebuilder-declarative-pattern/pkg/patterns/declarative/pkg/manifest"
+)
+
+func transformApplication(ctx context.Context, instance DeclarativeObject, objects *manifest.Objects, labelMaker LabelMaker) error {
+	log := log.Log
+
+	var app *manifest.Object
+	for _, o := range objects.Items {
```
Comment (justinsb) [inline pkg/patterns/declarative/application.go]: This looks similar to TransformApplicationFromStatus.  Should we extract a common function to "get the application from the manifest" (for follow-up PR).

**Comment by justinsb** [comment]
Looks like it'll need a rebase, and some things I suggest you follow-up on, but LGTM

/approve
/lgtm

File: examples/dashboard-operator/channels/packages/dashboard/1.8.3/manifest.yaml (hunk 1 above)
Comment (johnsonj) [reply examples/dashboard-operator/channels/packages/dashboard/1.8.3/manifest.yaml]: excellent point, i'm dropping my name as well :)

File: pkg/patterns/addon/application.go (hunk 2 above)
Comment (johnsonj) [reply pkg/patterns/addon/application.go]: done

File: pkg/patterns/addon/application.go (hunk 3 above)
Comment (johnsonj) [reply pkg/patterns/addon/application.go]: done

File: pkg/patterns/declarative/application.go (hunk 5 above)
Comment (johnsonj) [reply pkg/patterns/declarative/application.go]: done!

**Comment by justinsb** [comment]
Nice!

/approve
/lgtm

**Pull request merged**